src/hotspot/cpu/x86/templateTable_x86.cpp

```diff
@@ -3060,81 +3121,73 @@ void TemplateTable::jvmti_post_field_mod(Register cache, Register index, bool is
       __ cmovptr(Assembler::equal,
-                 c_rarg1, at_tos_p2()); // dtos (two word jvalue)
+                 obj, at_tos_p2()); // dtos (two word jvalue)
 #else
-      __ cmpl(rcx, ltos);
+      __ mov(obj, rsp);
+      __ cmpl(value, ltos);
       __ jccb(Assembler::equal, two_word);
-      __ cmpl(rcx, dtos);
+      __ cmpl(value, dtos);
       __ jccb(Assembler::equal, two_word);
-      __ addptr(rbx, Interpreter::expr_offset_in_bytes(1)); // one word jvalue (not ltos, dtos)
+      __ addptr(obj, Interpreter::expr_offset_in_bytes(1)); // one word jvalue (not ltos, dtos)
       __ jmpb(valsize_known);
 
       __ bind(two_word);
-      __ addptr(rbx, Interpreter::expr_offset_in_bytes(2)); // two words jvalue
+      __ addptr(obj, Interpreter::expr_offset_in_bytes(2)); // two words jvalue
 
       __ bind(valsize_known);
       // setup object pointer
-      __ movptr(rbx, Address(rbx, 0));
+      __ movptr(obj, Address(obj, 0));
 #endif
     }
-    // cache entry pointer
-    __ addptr(robj, in_bytes(cp_base_offset));
-    __ shll(RDX, LogBytesPerWord);
-    __ addptr(robj, RDX);
+
     // object (tos)
-    __ mov(RCX, rsp);
-    // c_rarg1: object pointer set up above (null if static)
-    // c_rarg2: cache entry pointer
-    // c_rarg3: jvalue object on the stack
+    __ mov(value, rsp);
+    // obj: object pointer set up above (null if static)
+    // cache: field entry pointer
+    // value: jvalue object on the stack
     __ call_VM(noreg,
-               CAST_FROM_FN_PTR(address,
-                                InterpreterRuntime::post_field_modification),
-               RBX, robj, RCX);
-    __ get_cache_and_index_at_bcp(cache, index, 1);
+              CAST_FROM_FN_PTR(address,
+                              InterpreterRuntime::post_field_modification),
+              obj, entry, value);
+    // Reload field entry
+    __ load_field_entry(cache, index);
     __ bind(L1);
   }
 }
 
 void TemplateTable::putfield_or_static(int byte_no, bool is_static, RewriteControl rc) {
   transition(vtos, vtos);
 
+  const Register obj = rcx;
   const Register cache = rcx;
   const Register index = rdx;
-  const Register obj   = rcx;
   const Register off   = rbx;
   const Register flags = rax;
 
-  resolve_cache_and_index(byte_no, cache, index, sizeof(u2));
+  resolve_cache_and_index_for_field(byte_no, cache, index);
   jvmti_post_field_mod(cache, index, is_static);
-  load_field_cp_cache_entry(obj, cache, index, off, flags, is_static);
+  load_resolved_field_entry(obj, cache, index, off, flags, is_static);
+  __ mov(flags, index);
 
   // [jk] not needed currently
   // volatile_barrier(Assembler::Membar_mask_bits(Assembler::LoadStore |
   //                                              Assembler::StoreStore));
 
   Label notVolatile, Done;
-  __ movl(rdx, flags);
-  __ shrl(rdx, ConstantPoolCacheEntry::is_volatile_shift);
-  __ andl(rdx, 0x1);
 
+  // Swap registers so flags has TOS state
+  __ xchgl(index, flags);
```

Comment:
I don't get this swap operation. Line 3178, index (containing the TosState) has been copied to flags (rax). The real flags for this field have been erased, and the volatile test below is not applied to the real flags but to the ToSState value.